### PR TITLE
Zif::UI::Input handling key inputs

### PR DESCRIPTION
**Pull request by davebytemfg**
I created a second branch with the borders and clickable stuff removed based on the discord discussion.

### CODE REVIEW

File: app/lib/zif/services/input_service.rb [hunk 1]
```diff
@@ -123,6 +139,33 @@ def process_click
...
 
+      # @api private
+      def process_key_event
+        return if @key_pressables.empty?
+
+        key_down = $gtk.args.inputs.keyboard.key_down.truthy_keys
+        # sometimes it's [:char, :raw_key, :s, :shift]
+        # sometimes it's [:char, :raw_key, :shift, :d]
+        # sometimes it's [:char, :raw_key, :shift_left, :d, :control_left, :control]
+        # so let's not care about shifts
```
Comment (danhealy) [inline app/lib/zif/services/input_service.rb:150]: Can you please elaborate on "sometimes" in these comments?

File: app/lib/zif/services/input_service.rb [hunk 2]
```diff
@@ -123,6 +139,33 @@ def process_click
...
+        key_down.delete(:shift)
+        return if key_down.empty?
+
+        if key_down == %i[char raw_key]
+          # this is a repeat char, aaaaaa, leave @last_key alone
+        elsif key_down[0] == :char # [:char, :raw_key, :s]
+          @last_key = key_down[2]
+        elsif key_down[0] == :raw_key # [:raw_key, :shift_left]
+          @last_key = key_down[1]
+        end
```
Comment (danhealy) [inline app/lib/zif/services/input_service.rb:160]: - Replace noop if statement with nested if
- Replace first and last index gets with `first` and `last` - check if this is appropriate?
```suggestion
        if key_down != %i[char raw_key]
          if key_down.first == :char # [:char, :raw_key, :s]
            @last_key = key_down.last
          elsif key_down.first == :raw_key # [:raw_key, :shift_left]
            @last_key = key_down[1]
          end
        end
```

File: app/lib/zif/ui/input.rb [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+    #      ...
+    #
+    class Input < Label
+      include Zif::KeyPressable
+      include Zif::Serializable
+
+      attr_rect
+      attr_accessor :max_length
+      attr_accessor :desired_keys, :special_keys, :map_keys
+      attr_accessor :has_focus
```
Comment (danhealy) [inline app/lib/zif/ui/input.rb:20]: These need to be split into their own lines and documented, see `clickable.rb` for an example

Additionally each method here should probably be marked `# @api private`

File: app/lib/zif/ui/input.rb [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+
+      @desired_keys = nil
+      @special_keys = nil
+      @map_keys     = nil
```
Comment (danhealy) [inline app/lib/zif/ui/input.rb:24]:
```suggestion
```

File: app/lib/zif/ui/input.rb [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+        # https://github.com/DragonRuby/dragonruby-game-toolkit-contrib/issues/78
+        super(text, size: size, alignment: alignment, font: font, ellipsis: ellipsis, r: r, g: g, b: b, a: a)
+
+        default_key_mappings
+        @on_key_down = ->(key) { handle_input(key) }
+      end
+
+      def default_key_mappings
+        @desired_keys = DEFAULT_DESIRED_KEYS
+        @special_keys = DEFAULT_SPECIAL_KEYS
```
Comment (danhealy) [inline app/lib/zif/ui/input.rb:54]: 👍

File: app/scenes/ui_sample.rb [hunk 6]
```diff
@@ -260,6 +270,7 @@ def perform_tick
       update_glass_panel
       update_progress_bar
       update_interactable_button
+      $gtk.args.outputs.labels << [@input] # you have to add the input on each tick to track changes
```
Comment (danhealy) [inline app/scenes/ui_sample.rb:273]: Can this be added as a `static_label`?

**Review by danhealy — CHANGES_REQUESTED** [review]
Looks pretty good!  The big change requested is the logic in `process_key_event`.  The comments and documentation remarks are important but I can take care of them

File: app/lib/zif/services/input_service.rb [hunk 7]
```diff
@@ -123,6 +139,24 @@ def process_click
       end
       # rubocop:enable Metrics/PerceivedComplexity
 
+      # @api private
+      def process_key_event
+        return if @key_pressables.empty?
+
+        keys_down = $gtk.args.inputs.text
+        $gtk.args.inputs.keyboard.key_down.truthy_keys.each do |k|
+          keys_down << k if Zif::UI::Input::DEFAULT_SPECIAL_KEYS.include?(k)
```
Comment (davebytemfg) [inline app/lib/zif/services/input_service.rb:148]: this is wrong to be using the default, right now I'm not sure how to figure out which keys are convertable to strings and can be added to the label and which ones are control characters. Might just make another list of control characters... TBD.

File: app/lib/zif/ui/input.rb [hunk 8]
```diff
@@ -0,0 +1,78 @@
...
+      ].freeze
+
+      DEFAULT_SPECIAL_KEYS = [
+        :backspace, :delete
+      ].freeze
+
+      def initialize(text='', size: -1, alignment: :left, font: 'font.tff', ellipsis: '…', r: 51, g: 51, b: 51, a: 255)
+        # super # calling super without args should have worked and passed the args up, but apparently only text??
+        # https://github.com/DragonRuby/dragonruby-game-toolkit-contrib/issues/78
+        super(text, size: size, alignment: alignment, font: font, ellipsis: ellipsis, r: r, g: g, b: b, a: a)
```
Comment (davebytemfg) [inline app/lib/zif/ui/input.rb:47]: Need to look at DR 2.12, do we also need to pass the new vertical_align attribute along?

**Pull request merged**